### PR TITLE
Plugins: Example CLI Plugin, a Resource Lister

### PR DESCRIPTION
**Pull request by clintoncwolfe**
Included in this PR is an example of Plugins v2 CliCommand plugin. It is very simple - it simply lists the names of the resources that InSpec knows about.

The plugin is exhaustively commented for pedagogic purposes.

The kit includes:
* an implementation of an InSpec CLI Command, using the InSpec PluginV2 API
* documentation in the README
* tests, at the unit and functional level
* a .gemspec, for packaging and publishing it as a gem
* a Gemfile, for managing its dependencies
* a Rakefile, for running development tasks
* Rubocop linting support for using the base InSpec project rubocop.yml (See Rakefile)

Additionally, there are two isolated commits that impact InSpec, outside of the example, both providing better support for an independent project to rely on the inspec installation for development support. [5188182](https://github.com/inspec/inspec/pull/3421/commits/51881823e57ac0ef2ed83b199e61e5cc80b15558) is straightforward, and [fff30db](https://github.com/inspec/inspec/pull/3421/commits/fff30db206ffd8c4cc9b8cca4f44ba9da499c4f9) is hair-raising.  Should either prove to be controversial, they can be cherry-picked into a separate PR.

I believe we should publish this example plugin to rubygems, after approval.

This PR supercedes #3400.

### CODE REVIEW

**Review by jquick — APPROVED** [review]
Thanks @clintoncwolfe. Lots of good info here. Very nice job explaining every part of the plugin.

File: .rubocop.yml [hunk 1]
```diff
@@ -5,7 +5,13 @@ AllCops:
   - Gemfile
   - Rakefile
   - 'test/**/*'
-  - 'examples/**/*'
+  # This is delicate; we want to include examples/plugins/*/lib
```
Comment (jerryaldrichiii) [inline .rubocop.yml:9]: I'm not following this comment. The wildcards below seem to have nothing to do with `examples/plugins/*/lib`

File: examples/plugins/inspec-resource-lister/README.md [hunk 2]
```diff
@@ -0,0 +1,62 @@
+# InSpec Plugin Example - Resource Lister
+
+This plugin provides an example of building a plugin for use with [InSpec](https://inspec.io). Its functionality is simple, but useful: list resources included with InSpec.
+
+## To Install this as a User
+
+You will need InSpec v2.3 or later.
+
+If you just want to use this (not learn how to write a plugin), you can so by simply running:
```
Comment (jerryaldrichiii) [inline examples/plugins/inspec-resource-lister/README.md]: Should this be `you can do so`? Also, I would reword to `just use this` and or `(and not learn how to write a plugin)`, but I'll leave those to you.

File: examples/plugins/inspec-resource-lister/README.md [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+yumrepo
+zfs_dataset
+zfs_pool
+------------------------------
+160 resources total
+```
+
+## Features of This Example Kit
+
+This example plugin is a full-fledged plugin example, with everything a real-world, industrial grade plugin would have, including:
```
Comment (jerryaldrichiii) [inline examples/plugins/inspec-resource-lister/README.md:35]: Industrial grade...heh :smile:

File: examples/plugins/inspec-resource-lister/README.md [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+* a .gemspec, for packaging and publishing it as a gem
+* a Gemfile, for managing its dependencies
+* a Rakefile, for running development tasks
+* Rubocop linting support for using the base InSpec project rubocop.yml (See Rakefile)
+
+You are encouraged to use this plugin as a starting point for real plugins.
+
+## Development of a Plugin
+
+[Plugin Development](https://github.com/inspec/inspec/blob/master/docs/dev/plugins.md) is documented on the inspec project on GitHub.  Additionally, this example
```
Comment (jerryaldrichiii) [inline examples/plugins/inspec-resource-lister/README.md]: `InSpec`?

File: examples/plugins/inspec-resource-lister/Rakefile [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+  t.warning = false
+end
+
+#------------------------------------------------------------------#
+#                    Code Style Tasks
+#------------------------------------------------------------------#
+require 'rubocop/rake_task'
+
+RuboCop::RakeTask.new(:lint) do |t|
+  # Choices of rubocop rules to enforce are deeply personal.
```
Comment (jerryaldrichiii) [inline examples/plugins/inspec-resource-lister/Rakefile:32]: I think it's `RuboCop`...but I'm not sure if we follow that internally.

File: examples/plugins/inspec-resource-lister/Rakefile [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+  # Here, we set things up so that your plugin will use the Bundler-installed
+  # inspec gem's copy of the InSpec project's rubocop.yml file (which
```
Comment (jerryaldrichiii) [inline examples/plugins/inspec-resource-lister/Rakefile:34]: `InSpec`?

File: examples/plugins/inspec-resource-lister/inspec-resource-lister.gemspec [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+Gem::Specification.new do |spec|
+  # Importantly, all InSpec plugins must be prefixed with `inspec-` (most
+  # plugins) or `train-` (plugins which add new connectivity features).
+  spec.name          = 'inspec-resource-lister'
+
+  # It is polite to namespace your plugin under InspecPlugins::YourPluginInCamelCase
+  spec.version       = InspecPlugins::ResourceLister::VERSION
+  spec.authors       = ['Clinton Wolfe']
+  spec.email         = ['cwolfe@chef.io']
+  spec.summary       = 'InSpec Plugin example, lists available resources'
```
Comment (jerryaldrichiii) [inline examples/plugins/inspec-resource-lister/inspec-resource-lister.gemspec:22]: Capital `Plugin`?

File: examples/plugins/inspec-resource-lister/inspec-resource-lister.gemspec [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+  spec.description   = 'Example for implementing an InSpec plugin.  This simply lists available resources.'
+  spec.homepage      = 'https://github.com/inspec/inspec/tree/master/examples/plugin'
+  spec.license       = 'Apache-2.0'
+
+  # Though complicated-looking, this is pretty standard for a gemspec.
```
Comment (jerryaldrichiii) [inline examples/plugins/inspec-resource-lister/inspec-resource-lister.gemspec:27]: Hyphens confuse me...this could be correct though.

File: examples/plugins/inspec-resource-lister/inspec-resource-lister.gemspec [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+    'lib/**/*', File::FNM_DOTMATCH
+  ).reject { |f| File.directory?(f) }
+  spec.require_paths = ['lib']
+
+  # If you rely on any other gems, list them here with any constraints.
+  # This is how `inspec plugin install` is able to manage your dependencies.
+  # For example, perhaps you are writing a thing that talks to AWS, and you
+  # want to ensure you have `aws-sdk` in a certain version.
+
+  # All plugins should mention inspec, > 2.2.78
```
Comment (jerryaldrichiii) [inline examples/plugins/inspec-resource-lister/inspec-resource-lister.gemspec:42]: Could we check for this in code?

File: examples/plugins/inspec-resource-lister/lib/inspec-resource-lister/plugin.rb [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+      # Notice that you can define multiple hooks with different names, and they
+      # don't have to match the plugin name.
+
+      # We'd like this to be list-resources, but Thor does not support hyphens
+      # see https://github.com/erikhuda/thor/pull/613
+      cli_command :listresources do
+        # Calling this hook doesn't mean list-resources is being executed - just
+        # that we should be ready to do so. So, load the file that defines the
+        # functionality.
+        # For example, inspec will activate this hook when `inspec help` is
```
Comment (jerryaldrichiii) [inline examples/plugins/inspec-resource-lister/lib/inspec-resource-lister/plugin.rb]: `InSpec`

File: examples/plugins/inspec-resource-lister/test/functional/inspec_resource_lister_test.rb [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+
+  # When thinking up scenarios to test, start with the simplest.
+  # Then think of each major feature, and exercise them.
+  # Running combinations of features makes sense if it is very likely,
+  # or a difficult / dangerous case.  You can always add more tests
+  # here as users find subtle problems.  In fact, having a user submit
+  # a PR that creates a failing functional test is a great way to
+  # capture the reproduction case.
+  # The simplest case:
+  describe "when run without an argument" do
```
Comment (jerryaldrichiii) [inline examples/plugins/inspec-resource-lister/test/functional/inspec_resource_lister_test.rb:26]: RuboCop would complain about double quotes without interpolation. The parent `describe` block is correct...but the rest seem to use double quotes. I can ignore it, but I'd like to see the same used within this file.

File: examples/plugins/inspec-resource-lister/test/functional/inspec_resource_lister_test.rb [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+    it "should mention the summary" do
+      outcome.stdout.must_include('resources total')
+    end
+  end
+
+  # Test the search pattern feature, in a couple of ways.
+  describe "when run with a search pattern that matches things" do
+    # Notice that the command line is changed here:
+    # "list all resources that have the word user in them"
+    let (:outcome) { run_inspec_process_with_this_plugin('listresources core user') }
```
Comment (jerryaldrichiii) [inline examples/plugins/inspec-resource-lister/test/functional/inspec_resource_lister_test.rb:59]: Should this be `let(:outcome)`

File: examples/plugins/inspec-resource-lister/test/functional/inspec_resource_lister_test.rb [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+        outcome.stdout.must_include(resource_name)
+      end
+    end
+    ['process', 'service', 'file'].each do |resource_name|
+      it "should NOT mention the '#{resource_name}' resource" do
+        outcome.stdout.wont_include(resource_name)
+      end
+    end
+  end
+  describe "when run with a search pattern that matches nothing" do
```
Comment (jerryaldrichiii) [inline examples/plugins/inspec-resource-lister/test/functional/inspec_resource_lister_test.rb:77]: There are newlines between blocks elsewhere in the file. I'll leave it to you though.

File: examples/plugins/inspec-resource-lister/test/functional/inspec_resource_lister_test.rb [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+    # Check for the summary
+    it "should mention a zero-resource summary" do
+      outcome.stdout.must_include('0 resources total')
+    end
+  end
+
+  # Exercise the summary option, which defaults to 'true'.
+  describe "when run with the no-summary flag" do
+    # Alter the command string to include the no-summary option
+    let (:outcome) { run_inspec_process_with_this_plugin('listresources core --no-summary') }
```
Comment (jerryaldrichiii) [inline examples/plugins/inspec-resource-lister/test/functional/inspec_resource_lister_test.rb]: Should this be `let(:outcome)`?

File: examples/plugins/inspec-resource-lister/test/helper.rb [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+
+# This file's job is to collect any libraries needed for testing, as well as provide
+# any utilities to make testing a plugin easier.
+
+# InSpec core provides a number of such libraries and facilities, in the file
+# lib/pligins/shared/core_plugin_test_helper.rb . So, one job in this file is
+# to locate and load that file.
+require 'inspec/../plugins/shared/core_plugin_test_helper'
+
+# Also load the Inspec plugin system. We need this so we can unit-test the plugin
```
Comment (jerryaldrichiii) [inline examples/plugins/inspec-resource-lister/test/helper.rb]: Should be `InSpec`

File: examples/plugins/inspec-resource-lister/test/helper.rb [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+# classes, which will rely on the plugin system.
+require 'inspec/plugin/v2'
+
+# Caution: loading all of InSpec (ie require 'inspec') may cause interference with
```
Comment (jerryaldrichiii) [inline examples/plugins/inspec-resource-lister/test/helper.rb]: Should be `i.e.`

File: examples/plugins/inspec-resource-lister/test/helper.rb [hunk 17]
```diff
@@ -0,0 +1,26 @@
...
+# minitest/spec; one symptom would be appearring to have no tests.
```
Comment (jerryaldrichiii) [inline examples/plugins/inspec-resource-lister/test/helper.rb]: Should be `appearing`

File: examples/plugins/inspec-resource-lister/test/helper.rb [hunk 18]
```diff
@@ -0,0 +1,26 @@
...
+# See https://github.com/inspec/inspec/issues/3380
+
+# You can select from a number of test harnesses.  Since InSpec uses Spec-style controls
+# in profile code, you will probably want to use something like minitest/spec, which provides
+# Spec-style tests.
+require 'minitest/spec'
+require 'minitest/autorun'
+
+# You might want to put some debugging tools here.  We run tests to find bugs, after all.
```
Comment (jerryaldrichiii) [inline examples/plugins/inspec-resource-lister/test/helper.rb:25]: Hmmm...I disagree with this. I don't think we should commit requires for `pry/byebug`/

File: examples/plugins/inspec-resource-lister/test/unit/README.md [hunk 19]
```diff
@@ -0,0 +1,17 @@
+# Unit Testing Area for Example Plugins
+
+## What example tests are provided?
+
+Here, since this is a CliCommand plugin, we provide two sets of unit tests:
+
+ * plugin_def_test.rb - Would be useful in any plugin.  Verifies that the plugin is properly detected and registered.
+ * cli_args_test.rb - verifies that the expected command are present, and that they have the expected options and args.
```
Comment (jerryaldrichiii) [inline examples/plugins/inspec-resource-lister/test/unit/README.md]: Looks like you capitalize after the hyphen in the sentence above, could we here too? Also, should be `commands are`.

File: examples/plugins/inspec-resource-lister/test/unit/README.md [hunk 20]
```diff
@@ -0,0 +1,17 @@
...
+
+## What are unit tests?
```
Comment (jerryaldrichiii) [inline examples/plugins/inspec-resource-lister/test/unit/README.md]: Seems capitalization varies throughout the docs here. I was expecting `What are Unit Tests?`.

File: examples/plugins/inspec-resource-lister/test/unit/cli_args_test.rb [hunk 21]
```diff
@@ -0,0 +1,64 @@
...
+require 'inspec-resource-lister/cli_command'
+
+# Because InSpec is a Spec-style test suite, we're going to use MiniTest::Spec
+# here, for familiar look and feel. However, this isn't InSpec (or RSpec) code.
+describe InspecPlugins::ResourceLister::CliCommand do
+
+  # When writing tests, you can use `let` to create variables that you
+  # can reference easily.
+
+  # This is the CLI COmmand implementation class.
```
Comment (jerryaldrichiii) [inline examples/plugins/inspec-resource-lister/test/unit/cli_args_test.rb]: Should be `command`? Or, maybe `Command`? Seems to vary in other docs.

File: examples/plugins/inspec-resource-lister/test/unit/cli_args_test.rb [hunk 22]
```diff
@@ -0,0 +1,64 @@
...
+  # To group tests together, you can nest 'describe' in minitest/spec
+  # (that is discouraged in InSpec control code.)
+  describe 'the core command' do
+
+    # Some tests through here use minitest Expectations, which attach to all
+    # Objects, and begin with 'must' (positive) or 'wont' (negative)
+    # See https://ruby-doc.org/stdlib-2.1.0/libdoc/minitest/rdoc/MiniTest/Expectations.html
+
+    # Option count OK?
+    it "should take one option" do
```
Comment (jerryaldrichiii) [inline examples/plugins/inspec-resource-lister/test/unit/cli_args_test.rb:34]: Also mixed usage of quote styles in this file.

File: lib/plugins/shared/core_plugin_test_helper.rb [hunk 23]
```diff
@@ -42,9 +44,57 @@ def run_inspec_process(command_line, opts = {})
...
+
+    Dir.mktmpdir do |tmp_dir|
+      File.write(File.join(tmp_dir, 'plugins.json'), JSON.generate(plugin_file_data))
+      opts[:env] ||= {}
+      opts[:env]['INSPEC_CONFIG_DIR'] = tmp_dir
+      run_inspec_process(command_line, opts)
+    end
+  end
+
+  def __find_plugin_path_from_caller(frames_back = 2)
```
Comment (jerryaldrichiii) [inline lib/plugins/shared/core_plugin_test_helper.rb]: Muhammad Ali level fancy footwork going on here.

**Review by jerryaldrichiii — CHANGES_REQUESTED** [review]
Great work @clintoncwolfe.  There is a superb amount of documentation here.

As usual, I only have grammatical/style requests. I'm also a bit troubled by the line breaks in comments not being consistent...but I won't hold up the PR over it.

File: .rubocop.yml (hunk 1 above)
Comment (jquick) [reply .rubocop.yml:9]: Right we are adding everything BUT `examples/plugins/*/lib` to the Exclude list here.

Comment (jerryaldrichiii) [reply .rubocop.yml:9]: :man_facepalming:

File: examples/plugins/inspec-resource-lister/README.md (hunk 4 above)
Comment (clintoncwolfe) [reply examples/plugins/inspec-resource-lister/README.md]: Refers to the identifier of the repo, not the name of the product.  I'll quote it.

File: examples/plugins/inspec-resource-lister/Rakefile (hunk 6 above)
Comment (clintoncwolfe) [reply examples/plugins/inspec-resource-lister/Rakefile:34]: No, this refers to a gem name, which is literally 'inspec'.

File: examples/plugins/inspec-resource-lister/inspec-resource-lister.gemspec (hunk 9 above)
Comment (clintoncwolfe) [reply examples/plugins/inspec-resource-lister/inspec-resource-lister.gemspec:42]: This is a template / example.  We will have no code exposure to it until they try to run it against inspec.

If they used too low of a version (which we cannot go back in time and add checks for) it will fail because they are trying to use a plugin system that does not exist.

File: examples/plugins/inspec-resource-lister/test/functional/inspec_resource_lister_test.rb (hunk 11 above)
Comment (clintoncwolfe) [reply examples/plugins/inspec-resource-lister/test/functional/inspec_resource_lister_test.rb:26]: Please use the team's agreed-upon guidelines for where and how we apply style rules.

File: examples/plugins/inspec-resource-lister/test/helper.rb (hunk 18 above)
Comment (clintoncwolfe) [reply examples/plugins/inspec-resource-lister/test/helper.rb:25]: Please bring up an argument for your position in the next standup.

**Review by jerryaldrichiii — APPROVED** [review]
